packages/test-kit/src/sync-utils-contract.ts

```diff
@@ -0,0 +1,63 @@
+import { expect } from 'chai';
+import type { IFileSystemSync } from '@file-services/types';
+import type { ITestInput } from './types';
+
+export function syncUtilsContract(testProvider: () => Promise<ITestInput<IFileSystemSync>>): void {
+  describe('SYNC utilities contract', () => {
+    let testInput: ITestInput<IFileSystemSync>;
+
+    beforeEach(async () => (testInput = await testProvider()));
+    afterEach(async () => await testInput.dispose());
+
+    describe('ensureDirectorySync', () => {
+      it(`creates intermediate directories`, () => {
+        const { fs, tempDirectoryPath } = testInput;
+
+        fs.populateDirectorySync(tempDirectoryPath, { a: {} });
+        const newDirPath = fs.join(tempDirectoryPath, 'animals', 'mammals', 'chiroptera');
+        fs.ensureDirectorySync(newDirPath);
+
+        expect(fs.directoryExistsSync(newDirPath)).to.equal(true);
+      });
+
+      it(`succeeds if directory already exists, preserves its contents`, () => {
+        const { fs, tempDirectoryPath } = testInput;
+        fs.populateDirectorySync(tempDirectoryPath, { animals: { mammals: { 'vesper-bat.txt': '🦇' } } });
+        fs.ensureDirectorySync(fs.join(tempDirectoryPath, 'animals', 'mammals'));
+
+        const filePath = fs.join(tempDirectoryPath, 'animals', 'mammals', 'vesper-bat.txt');
+        expect(fs.readFileSync(filePath, 'utf8')).to.equal('🦇');
+      });
+
+      it(`throws when attempting to overwrite existing file`, () => {
+        const { fs, tempDirectoryPath } = testInput;
+
+        fs.populateDirectorySync(tempDirectoryPath, { 'vesper-bat.txt': '🦇' });
+        const newDirPath = fs.join(tempDirectoryPath, 'vesper-bat.txt');
+
+        expect(() => fs.ensureDirectorySync(newDirPath)).to.throw('EEXIST');
+      });
+
+      it(`throws when attempting to create a directory inside of a file`, () => {
+        const { fs, tempDirectoryPath } = testInput;
+
+        fs.populateDirectorySync(tempDirectoryPath, { 'vesper-bat.txt': '🦇' });
+        const newDirPath = fs.join(tempDirectoryPath, 'vesper-bat.txt', 'habitat');
+
+        expect(() => fs.ensureDirectorySync(newDirPath)).to.throw('ENOTDIR');
+      });
+
+      it('handles special paths gracefully', () => {
+        const { fs, tempDirectoryPath } = testInput;
+
+        fs.chdir(tempDirectoryPath);
```

Comment:
that's why other tests failed. test changed the cwd for the entire process. After this happened, every `nodeFs.resolve` (=== native `path.resolve`) calls used this `tempDirectory` as the `cwd`.

if you delete the folder that process.cwd() points to, both `process.cwd()` and `path.resolve(...)` will throw when called.

the tempDirectoryPath is deleted after the contract finishes, hence the exception.

long story short: is there a way to test this without having to call chdir? or do we want to revert the chdir after the test?

maybe:
```
try {
// test code
} finally{
// restore chdir
}
```
without a catch, any exception will propagate to the test runner, as it should.

Reply:
OMG, you're probably right. What do you think about restoring cwd after each test?

Reply:
Yep, you were right.